go/tasks/plugins/k8s/kfoperators/mpi/mpi.go

```diff
@@ -0,0 +1,150 @@
+	return &mpi.MPIJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       mpi.Kind,
+			APIVersion: mpi.SchemeGroupVersion.String(),
+		},
+	}, nil
+}
+
+// Defines a func to create the full resource object that will be posted to k8s.
+func (mpiOperatorResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (client.Object, error) {
+	taskTemplate, err := taskCtx.TaskReader().Read(ctx)
+
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "unable to fetch task specification [%v]", err.Error())
+	} else if taskTemplate == nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "nil task specification")
+	}
+
+	mpiTaskExtraArgs := plugins.DistributedMPITrainingTask{}
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &mpiTaskExtraArgs)
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "invalid TaskSpecification [%v], Err: [%v]", taskTemplate.GetCustom(), err.Error())
+	}
+
+	podSpec, err := flytek8s.ToK8sPodSpec(ctx, taskCtx)
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "Unable to create pod spec: [%v]", err.Error())
+	}
+	workers := mpiTaskExtraArgs.GetNumWorkers()
+	launcherReplicas := mpiTaskExtraArgs.GetNumLauncherReplicas()
+	slots := mpiTaskExtraArgs.GetSlots()
+
+	workersPodSpec := podSpec.DeepCopy()
+	for k := range workersPodSpec.Containers {
+		workersPodSpec.Containers[k].Args = []string{}
+		workersPodSpec.Containers[k].Command = []string{}
+	}
+
+	jobSpec := mpi.MPIJobSpec{
+		SlotsPerWorker: &slots,
+		MPIReplicaSpecs: map[mpi.MPIReplicaType]*commonKf.ReplicaSpec{
+			mpi.MPIReplicaTypeLauncher: &commonKf.ReplicaSpec{
+				Replicas: &launcherReplicas,
+				Template: v1.PodTemplateSpec{
+					Spec: *podSpec,
+				},
+				RestartPolicy: commonKf.RestartPolicyNever,
+			},
+			mpi.MPIReplicaTypeWorker: &commonKf.ReplicaSpec{
+				Replicas: &workers,
+				Template: v1.PodTemplateSpec{
+					Spec: *workersPodSpec,
+				},
+				RestartPolicy: commonKf.RestartPolicyNever,
+			},
+		},
+	}
+
+	job := &mpi.MPIJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       mpi.Kind,
+			APIVersion: mpi.SchemeGroupVersion.String(),
+		},
+		Spec: jobSpec,
+	}
+
+	return job, nil
+}
+
+// Analyses the k8s resource and reports the status as TaskPhase. This call is expected to be relatively fast,
+// any operations that might take a long time (limits are configured system-wide) should be offloaded to the
+// background.
+func (mpiOperatorResourceHandler) GetTaskPhase(_ context.Context, pluginContext k8s.PluginContext, resource client.Object) (pluginsCore.PhaseInfo, error) {
+	app := resource.(*mpi.MPIJob)
+
+	numWorkers := app.Spec.MPIReplicaSpecs[mpi.MPIReplicaTypeWorker].Replicas
+	numLauncherReplicas := app.Spec.MPIReplicaSpecs[mpi.MPIReplicaTypeLauncher].Replicas
+
+	taskLogs, err := common.GetLogs(common.MPITaskType, app.Name, app.Namespace,
+		*numWorkers, *numLauncherReplicas, 0)
```

Comment:
do you need to defensively check whether `numWorkers` or `numLauncherReplicas` might be nil before dereferencing?